assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.test.js

```diff
@@ -0,0 +1,496 @@
+
+describe( 'SettingsEnhancedMeasurementSwitch', () => {
+	const { accounts } = fixtures.accountsPropertiesProfiles;
+	const { properties, webDataStreams } = ga4Fixtures;
+	const accountID = accounts[ 0 ].id;
+	const propertyID = properties[ 0 ]._id;
+	const webDataStreamID = webDataStreams[ 0 ]._id;
+
+	let enhancedMeasurementSettingsMock;
+	let registry;
+
+	function setupRegistry() {
+		enhancedMeasurementSettingsMock = {
+			fileDownloadsEnabled: null,
+			name: `properties/${ propertyID }/dataStreams/${ webDataStreamID }/enhancedMeasurementSettings`,
+			outboundClicksEnabled: null,
+			pageChangesEnabled: null,
+			scrollsEnabled: null,
+			searchQueryParameter: 'q,s,search,query,keyword',
+			siteSearchEnabled: null,
+			streamEnabled: true,
+			uriQueryParameter: null,
+			videoEngagementEnabled: null,
+		};
+
+		registry = createTestRegistry();
+
+		registry.dispatch( MODULES_ANALYTICS ).receiveGetSettings( {
+			accountID,
+		} );
+
+		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
+			propertyID,
+			webDataStreamID,
+		} );
+
+		registry.dispatch( MODULES_ANALYTICS ).receiveGetAccounts( accounts );
+		registry
+			.dispatch( MODULES_ANALYTICS )
+			.finishResolution( 'getAccounts', [] );
+
+		registry
+			.dispatch( MODULES_ANALYTICS_4 )
+			.receiveGetProperties( properties, { accountID } );
+		registry
+			.dispatch( MODULES_ANALYTICS_4 )
+			.finishResolution( 'getProperties', [ accountID ] );
+
+		registry
+			.dispatch( MODULES_ANALYTICS_4 )
+			.receiveGetWebDataStreams( webDataStreams, {
+				propertyID,
+			} );
+		registry
+			.dispatch( MODULES_ANALYTICS_4 )
+			.finishResolution( 'getWebDataStreams', [ propertyID ] );
+	}
+
+	beforeEach( () => {
+		setupRegistry();
+
+		registry
+			.dispatch( MODULES_ANALYTICS_4 )
+			.receiveGetEnhancedMeasurementSettings(
+				enhancedMeasurementSettingsMock,
+				{ propertyID, webDataStreamID }
+			);
+		registry
+			.dispatch( MODULES_ANALYTICS_4 )
+			.finishResolution( 'getEnhancedMeasurementSettings', [
+				propertyID,
+				webDataStreamID,
+			] );
+	} );
+
+	describe.each( [
+		[ 'enabled', true ],
+		[ 'disabled', false ],
+	] )(
+		'when enhanced measurement is %s for the web data stream',
+		( position, streamEnabled ) => {
+			it( `should render correctly, with the switch defaulting to the ${ position } position`, async () => {
+				await registry
+					.dispatch( MODULES_ANALYTICS_4 )
+					.setEnhancedMeasurementStreamEnabled(
+						propertyID,
+						webDataStreamID,
+						streamEnabled
+					);
+
+				const { container, getByLabelText } = render(
+					<SettingsEnhancedMeasurementSwitch hasAnalytics4Access />,
+					{
+						registry,
+					}
+				);
+
+				expect( container ).toMatchSnapshot();
+
+				const switchControl = getByLabelText(
+					'Enable enhanced measurement'
+				);
+
+				if ( streamEnabled ) {
+					expect( switchControl ).toBeChecked();
+				} else {
+					expect( switchControl ).not.toBeChecked();
+				}
+			} );
+		}
+	);
```

Comment:
Seeing if statements in the tests makes them harder to read, as the reader has to backtrack above to make sense of the conditional logic here. I think this `.each` is saving minimal lines of code, but it's reducing the readability of these tests. In this case the test code isn't even shared because of this conditional.

I think it'd be better to write out each test spec without variables so the specs are written in plain English and the test code isn't conditional. Tests prioritising DRY principles reduce readability in exchange for conciseness, but for tests I think that's less useful.

Reply:
I'm generally in full agreement here, preferring DAMP over DRY in a test context. As I was writing this test, I thought this particular case was "acceptable", due to the simple nature of the test. However I am more than happy to unroll these tests, and have done so. :)